File: test/BridgeNetTests/TypeScript/App1/2_Interfaces.cs
namespace Interfaces
{
    //public struct Point
    //{
    //    public int X;
    //    public int Y;

    //    public Point(int x, int y)
    //    {
    //        this.X = x;
    //        this.Y = y;
    //    }
    //}

    public interface Interface1
    {
        int Property
        {
            get;
            set;
        }
    }

    public interface Interface2 : Interface1
    {
        void Method1();

        void Method2(string i);

        int Method3();

        bool Method4(Interface1 i);
    }

    public interface Interface3 : Interface2
    {
        Interface2 Method5(Interface3 i);
    }

    public interface Interface4
    {
        // #291
        void Method6(out bool b);

        void Method7(int i, out bool b);

        void Method8(ref string s);

        void Method9(int i, ref string s);

        void Method10(int i, ref bool b, ref string s);
    }

    public interface Interface6
    {
        int Property
        {
            get;
            set;
        }

        int GetProperty();

        void SetProperty(string s);

        void SetProperty(int i);
    }

    public interface Interface61
    {
        int Property
        {
            get;
            set;
        }
    }

    public interface Interface62
    {
        int GetProperty();

        void SetProperty(string s);

        void SetProperty(int i);
    }

    public class Class1 : Interface1
    {
        public int Field = 200;

        private int property = 100;

        public int Property
        {
            get
            {
                return this.property;
            }
            set
            {
                this.property = value;
            }
        }
    }

    public class Class2 : Class1, Interface2
    {
        public void Method1()
        {
            Field = 1;
            Property = 2;
        }

        public void Method2(string s)
        {
            Field = s.Length;
        }

        public int Method3()
        {
            return Field;
        }

        public bool Method4(Interface1 i)
        {
            Field = i.Property;

            return true;
        }
    }

    public class Class3 : Class2, Interface3
    {
        public Interface2 Method5(Interface3 i)
        {
            return i;
        }
    }

    public class Class4 : Interface4
    {
        public void Method6(out bool b)
        {
            b = true;
        }

        public void Method7(int i, out bool b)
        {
            b = true;
        }

        public void Method8(ref string s)
        {
            s = s + "Method8";
        }

        public void Method9(int i, ref string s)
        {
            s = s + i;
        }

        public void Method10(int i, ref bool b, ref string s)
        {
            b = true;
            s = s + i;
        }
    }

    public class Class6 : Interface6
    {
        public int Property
        {
            get;
            set;
        }

        public int MethodProperty
        {
            get;
            set;
        }

        public int GetProperty()
        {
            return this.MethodProperty;
        }

        public void SetProperty(string s)
        {
            this.MethodProperty = s.Length;
        }

        public void SetProperty(int i)
        {
            this.MethodProperty = i;
        }
    }

    // TODO Issue?
    // Cannot translate interface (interfaces.Interface61) member '[Property interfaces.Class63_1.Property]' in 'interfaces.Class63_1' due name conflicts. Please rename methods or refactor your code    at Bridge.Contract.OverloadsCollection.GetOverloadName(IMember definition) in c:\NET\Bridge\Builder\Contract\OverloadsCollection.cs:line 865

    //public class Class63_1 : Interface61, Interface62
    //{
    //    public int Property { get; set; }

    //    public int MethodProperty { get; set; }

    //    public int GetProperty()
    //    {
    //        return this.MethodProperty;
    //    }
    //    public void SetProperty(string s)
    //    {
    //        this.MethodProperty = s.Length;
    //    }
    //    public void SetProperty(int i)
    //    {
    //        this.MethodProperty = i;
    //    }
    //}
}

File: test/BridgeNetTests/TypeScript/Batch1/2_Interfaces.ts
/// <reference path="..\..\Runner\resources\qunit\qunit.d.ts" />
/// <reference path="..\..\Runner\TypeScript\App1\bridge.d.ts" />
/// <reference path="..\..\Runner\TypeScript\App1\Interfaces.d.ts" />

"use strict";

QUnit.module("TypeScript - Interfaces");
QUnit.test("Simple field and property", function (assert) {
    var instance: Interfaces.Class1 = new Interfaces.Class1();

    TestInstance1(assert, instance);
});

QUnit.test("Four methods", function (assert) {
    var instance: Interfaces.Class2 = new Interfaces.Class2();

    TestInstance1(assert, instance);

    TestInstance2(assert, instance);
});

QUnit.test("Fifth method through Interface3", function (assert) {
    var instance: Interfaces.Class3 = new Interfaces.Class3();

    TestInstance1(assert, instance);

    TestInstance2(assert, instance);

    TestInstance3(assert, instance);
});

QUnit.test("Out and reference parameters", function (assert) {
    var instance: Interfaces.Class4 = new Interfaces.Class4();
    TestInstance4(assert, instance);
});

QUnit.test("Property setter and getter with same names as methods", function (assert) {
    var instance: Interfaces.Class6 = new Interfaces.Class6();
    TestInstance6(assert, instance);
});

function TestInstance1(assert: QUnitAssert, instance: Interfaces.Class1) {
    var interface1: Interfaces.Interface1 = instance;

    assert.deepEqual(interface1.Interfaces$Interface1$Property, 100, "Interface1 Property getter");
    assert.deepEqual(instance.Field, 200, "Class1 Field initial value");

    interface1.Interfaces$Interface1$Property = 300;
    assert.deepEqual(instance.Property, 300, "Class1 Property setter");
}

function TestInstance2(assert: QUnitAssert, instance: Interfaces.Class2) {
    var interface2: Interfaces.Interface2 = instance;

    interface2.Interfaces$Interface2$Method1();
    assert.deepEqual(interface2.Interfaces$Interface1$Property, 2, "Method1() through Property");
    assert.deepEqual(instance.Field, 1, "Method1() through Fileld");

    interface2.Interfaces$Interface2$Method2("1234567");
    assert.deepEqual(instance.Field, 7, "Method2() through Field");

    assert.deepEqual(interface2.Interfaces$Interface2$Method3(), instance.Field, "Method3 through Field");

    assert.ok(interface2.Interfaces$Interface2$Method4(instance), "Method4 through return result");
    assert.deepEqual(instance.Field, interface2.Interfaces$Interface1$Property, "Method4 through Field");
}

function TestInstance3(assert: QUnitAssert, instance: Interfaces.Class3) {
    var interface3: Interfaces.Interface3 = instance;

    var interface2 = interface3.Interfaces$Interface3$Method5(interface3);
    assert.deepEqual(interface3.Interfaces$Interface1$Property, interface2.Interfaces$Interface1$Property, "Method5 through Property");

    var instance1: Interfaces.Class1 = instance;
    assert.deepEqual(instance1.Field, instance.Field, "Method5 through Field");
}

function TestInstance4(assert: QUnitAssert, instance: Interfaces.Class4) {
    var interface4: Interfaces.Interface4 = instance;

    var boolValue = false;
    var boolRef = { v: boolValue };

    instance.Method6(boolRef);
    assert.deepEqual(boolRef.v, true, "Method6() out bool parameter");

    boolRef.v = false;
    instance.Method7(1, boolRef);
    assert.deepEqual(boolRef.v, true, "Method7() int, out bool parameter");

    var stringValue = "ABC_";
    var stringRef = { v: stringValue };
    instance.Method8(stringRef);
    assert.deepEqual(stringRef.v, "ABC_Method8", "Method8() ref string parameter");

    stringRef.v = "ABC_";
    instance.Method9(1, stringRef);
    assert.deepEqual(stringRef.v, "ABC_1", "Method9() int, ref string parameter");

    stringRef.v = "ABC_";
    boolRef.v = false;
    instance.Method10(2, boolRef, stringRef);
    assert.deepEqual(boolRef.v, true, "Method10() int, ref bool, out bool parameter");
    assert.deepEqual(stringRef.v, "ABC_2", "Method10() int, ref bool, ref string parameter");
}

function TestInstance6(assert: QUnitAssert, instance: Interfaces.Class6) {
    var interface6: Interfaces.Interface6 = instance;

    interface6.Interfaces$Interface6$Property = 1;
    assert.deepEqual(instance.Property, 1, "Property getter and setter");

    interface6.Interfaces$Interface6$SetProperty$1("12");
    assert.deepEqual(instance.MethodProperty, 2, "setProperty$1(string) and MethodProperty");
    assert.deepEqual(instance.GetProperty(), 2, "setPropert$1(string) and getProperty");
    assert.notDeepEqual(instance.Property, instance.GetProperty(), "getProperty and getProperty$3");

    interface6.Interfaces$Interface6$SetProperty(3);
    assert.deepEqual(instance.MethodProperty, 3, "setProperty(int) and MethodProperty");
    assert.deepEqual(instance.GetProperty(), 3, "setPropert$1(int) and getProperty");
    assert.notDeepEqual(instance.Property, instance.GetProperty(), "getProperty$3 and getProperty");

    //var interface61: Interfaces.Interface61 = instance;
    //var interface62: Interfaces.Interface62 = instance;
}
